### ISSUE TITLE
Lint against `HashMap`s where the key is not `Eq + Hash`

### ISSUE THREAD

**Issue by cyrgani** [rust-lang/rust-clippy#12866]
### What it does

An empty `HashMap<K, V>` can be constructed with any type of key, but to actually allow inserting values into it, the key type must implement `Eq` and `Hash`.
This lint would warn if it encounters any type that does not implement `Eq + Hash` used as a `HashMap` key and, if possible, suggest deriving the missing traits (`PartialEq`, `Eq`, `Hash`).

### Advantage

- The `HashMap` is fully usable.

### Drawbacks

- There might be a legitimate use case for a `HashMap` without `K: Eq + Hash` (I don't know if there is one).

### Example

```rust
use std::collections::HashMap;
struct Key(u32);
struct Other(HashMap<Key, String>);
```

Could be written as:

```rust
use std::collections::HashMap;
#[derive(PartialEq, Eq, Hash)]
struct Key(u32);
struct Other(HashMap<Key, String>);
```

**Comment by Alexendoo** [comment]
Behind a feature gate there's [the raw entry API](https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#impl-HashMap%3CK,+V,+S%3E-2) which is useful without `K: Eq + Hash`, though the future of that API is unclear

Would the lint catch any issues would be my question, my impression is you'd run into an error when you go to use the map anyway that'll let you know it needs to be `Hash/Eq`

**Comment by cyrgani** [comment]
I have encountered this sometimes when I did not derive `Eq` or `Hash` for the key and was confused by rust-analyzer not suggesting methods like `insert` or `get`, but I understand why a separate lint for this might be unnecessary.
